Improve ICANN verification flow

mikeshelton1503 opened the pull request:
Fixes #2478 #13231 
Also as a bonus, added a fix for #6098

This improves the presentation of the ICANN email verification step.

Before | After
------- | -------
![image](https://user-images.githubusercontent.com/12596797/28026935-81c93576-6565-11e7-8f15-f66ef2a916b6.png) | ![image](https://user-images.githubusercontent.com/12596797/28090922-12ed912c-665c-11e7-8225-64393c74384d.png)
![image](https://user-images.githubusercontent.com/12596797/28026990-ae0cf00a-6565-11e7-8c42-35919a6c271d.png) | ![image](https://user-images.githubusercontent.com/12596797/28090954-291c4916-665c-11e7-8a26-d2a459b3afc2.png)




Changes:
- Updated copy for ICANN verification card - removed all mention of ICANN.
- ^^ That includes adding specific mention of what email address the email was and will be sent to and a direct link to change it.
- Replace global success notices with a status bar that communicates the action needed.


How to test:
- Purchase a domain but don't verify your email address yet.
- Go to Domains and click on the domain you just bought.
- Affirm you see the Email verification box at the top.
- Click "Send Again".
- Affirm you were shown a "Email sent" message.
- After 5 seconds the message should disappear. Then click "Change Email Address".
- Affirm you are taken to the Edit Contact Info screen.

mikeshelton1503 commented:
Also @ranh this could use a copy review.

michelleweber commented:
Just checking -- do you need a copy review from @ranh, or from Editorial? (The "Needs Copy Review" label pings us.)

mikeshelton1503 commented:
>Just checking -- do you need a copy review from @ranh, or from Editorial? (The "Needs Copy Review" label pings us.)

Oh I see! I intended it for Ran but won't turn down more copy reviews but probably not necessary for two reviews :)

klimeryk commented:
Awesome stuff, @mikeshelton1503! 💟 
I'll try to finish the developer part of this before the end of the week!

klimeryk commented:
A suggestion comes to mind - AFAIK, we shouldn't be using the global ("floating") notices for stuff like success notices. Maybe we could, for example, on success substitute the verification card with a nice, big green notice?

mikeshelton1503 commented:
>A suggestion comes to mind - AFAIK, we shouldn't be using the global ("floating") notices for stuff like success notices.

Oh, this would be news to me. Where did you hear this? Seems to me that a floating notice is the perfect type for a quick acknowledgement of an action such as success. I was out of Calypso land for much of last year so I could have missed it. cc @folletto @drw158 

I'm certainly not opposed to keeping the success acknowledgment inline within the verification card but wouldn't want to use a big green notice. It is a "successful" action but not a successful completion of the task since you need to check the email so we don't want to give the impression that you're done.

ranh commented:
I don't understand some of the changes you made to the copy. Can you explain why you changed the reason we give for the need to verify? I didn't know it was necessary to verify the contact information to make changes to the domain. Even if it is, most people will not want to make changes to their domain, so they may assume this message doesn't apply to them. The fact that the domain will stop working if not verified is probably going to be more relevant to most users. 

We can cut down on the term "verification". For example instead of "look for the verification message", we can say "look for our email", and instead of "resend verification email", we can just say "resend email".

I tend to agree with @klimeryk, I don't think the global notice is a good fit here. Is there any specific reason to have one part of the UI in one area of the screen and another part of the same UI in an entirely different area? 

I think we would be better off with a much more subtle indication, near the button itself.

ranh commented:
> I didn't know it was necessary to verify the contact information to make changes to the domain

I see now that this comes from #2478. And that I actually wrote the piece of copy in question 😂 

However, the copy in that issue was meant specifically for the name servers screen, not the main domain screen. If users are on the name servers screen, it does make sense to assume they are trying to make changes, but on the main domain screen that assumption doesn't hold. So I think we're better off focusing on the domain getting suspended there.

folletto commented:
> AFAIK, we shouldn't be using the global ("floating") notices for stuff like success notices.

The patterns is actually exactly that: floating notices are for any quick confirmation of an action that requires text and has no other in-screen UI change going on.

So yeah, I confirm:

> Seems to me that a floating notice is the perfect type for a quick acknowledgement of an action such as success.

That's the use of a global notice bubble.
Short text is great too, so good job in shortening it. Also good take in switching to the info notice as it requires a follow-up action to be completed.

mikeshelton1503 commented:
>Note however that if there's a way to message this inside the page UI in a different way, like changing status on some of the UI blocks on the page to say "Email sent, waiting for reply [resend]" or something, then the UI itself will communicate that and there would be no need for a bubble notification.

Well I think this is ideal though I'm not aware of any existing patterns for this. Are there any? Anyways I'll work on keeping it within the UI, I agree it makes the most sense here.

mikeshelton1503 commented:
>However, the copy in that issue was meant specifically for the name servers screen, not the main domain screen. 

Oh I see, that makes sense. I see now that is a separate step. There is also a version of the copy I edited that specifically for nameservers in the code. I updated that copy too.

Anyways I'll update the nameservers step with this copy and add new copy for the main screen unless you have suggestions.

folletto commented:
> Well I think this is ideal though I'm not aware of any existing patterns for this. Are there any? 

It's not a pattern: it depends on the UI you're transforming to the informative temporary state. 
I think one of the best and most explicit examples is Imports, like #10770. 

But you can also see our "Manage Purchases" screen as a UI that changes to different states depending on extra-UI factors, so that's another example probably.

I hope it helps :)

folletto commented:
> he meant that about the old

Ah yes, you got caught up in the period where we were transitioning that pattern. Sorry!

mikeshelton1503 commented:
Ok I've made some changes based on the above feedback. Instead of using a success notice bubble, I switched to an inline method using statuses. There are now two statuses: `waiting` and `email-sent`. I also changed up the copy a bit, letting the status convey the action and the intro copy explain the process.

When a user first visit domain section they will see this, which is the `waiting` status:

![image](https://user-images.githubusercontent.com/12596797/27981592-42fbab40-635c-11e7-869e-b2714c168665.png)

When a user clicks "Send Again" the button will go to disabled state and the button copy will change to "Sending ...".

![image](https://user-images.githubusercontent.com/12596797/27981622-b4aacc76-635c-11e7-8bf0-4db3033a23db.png)

Once the action is complete, the contents of the status bar fade out and a new status fades in. This is the `email-sent` status. It appears until the page is navigated away from or refreshed, then the status changes back to `waiting`.

![image](https://user-images.githubusercontent.com/12596797/27981639-fd22ae38-635c-11e7-8c5e-238cd65a7265.png)

I just pushed up changes to put this new structure in place but I'll need help with the functionality as described above. cc @klimeryk

ranh commented:
These notices look great, I really like how you incorporated the status indicator inside the notice. 

The big button for "send again" makes it seem like it's the obvious next action - is that intentional? If we want to make sure that users will have a new email from us and not have to dig through the archives for the one we sent after purchase, we can just resend the email automatically when we show the notice, rather than add an extra step for it.

This might be nitpicking, but I still think that users may end up clicking the "send again" button and thinking they were done. The green icon, disabled button and "email sent" text can be confusing, but even without them, just the fact that users "acted" on the notice in some way may be enough to have them skip the actual verification altogether. 

That's why it may be better to not have a prominent next action to do on this screen, so that it's very clear that the next action actually has to be done elsewhere.

klimeryk commented:
I like @ranh's suggestion about re-sending the email when we show that notice... but I think it will be actually harder to implement that approach properly than an explicit button. That's because how do we decide when to send the email? OK,  the user is shown the notice, we send the email, but then the user refreshes the view, navigates back and forward again, etc. We don't want to end up spamming the user, but don't want to have them guessing (or the HE helping them) either if the email has been re-sent. We could introduce some timeout before we re-send the email, but, for example, for OpenHRS domains, updating the email is a very quick action and the user might want to re-send the email ASAP.

Maybe we could instead hide the `Send Again` button when the user sends the email? That way, they will be more likely to read the copy on the left of it urging them to check their email and verify it. Of course, if they navigate away and come back, we'll show the button again.

klimeryk commented:
Also: cc @ehti, I'd love to hear your opinion about these improvements from HE point of view 🙇

klimeryk commented:
@mikeshelton1503 I've cleaned up the code a bit and done the developer tasks outlined in the first post. I've went ahead and implemented hiding the send button when the email has been sent - of course, it's easy to revert that if that idea is not to your or Ran's liking. I've also used the `busy` style for the button when it's submitting to better show some sort of indication that an action is performed.

ranh commented:
> re-sending the email when we show that notice... but I think it will be actually harder to implement that approach properly than an explicit button. That's because how do we decide when to send the email? 

Agreed, it's more complicated than it sounds at first, but the point is not to automate the process of resending the email, just to make it clear that the action required has to be done elsewhere, not on this screen. 

Starting the flow by clicking on "Resend Email" is a good first step because it ensures that the message will be (relatively) easy to find. But based on the experience from the current flow, we just need to be extra clear that this button is only the first step, not the last step.

folletto commented:
> Instead of using a success notice bubble, I switched to an inline method using statuses.

This is _really_ great. 
I think there's a whole lot more clarity, and far better in terms of flows.

Small nitpick: the design seem using a `$gray`. For text we should never use anything lighter than `$gray-text-min` ([ref](https://github.com/Automattic/wp-calypso/blob/a70587157b33783ba4d5e0e4a1e136680d459d51/assets/stylesheets/shared/_colors.scss#L18)).

> That's why it may be better to not have a prominent next action to do on this screen, so that it's very clear that the next action actually has to be done elsewhere.

This is a good point.
A simple way to address it would be to show the button as normal (white) instead of primary (blue), to give more visual weight to the text on the side. Another approach would be to turn away from a button and use a text like the line already there: 

> Sent to person@example.org
> Haven't received the email? _Send it again_ or _change your email address_.

Or something along these lines.

ehti commented:
Thanks @klimeryk .

Users reach out for suspended domains/email verification primarily for two reasons.

1. They didn't get the email. Often, this is because of a typo in their email OR email ends up in spam.

2. Site isn't working anymore. They don't really know it's because their email isn't verified. In a lot of cases this is related to no.1, i.e. didn't receive the email as well.

In the messaging above, it's GREAT that we're now mentioning the address where we're sending the emails. However, is it prominent enough to notice typos? Or, can there be a mention of something along the lines @folletto mentioned above, e.g.:

> Sent to person@example.org
> Haven't received the email? Check spam/junk or _send it again_.
> Is the email incorrect? Change it from here.

This covers:

1. Typos. This happens a lot.
2. Emails going to spam. This too happens at times.

Also, why are we saying:
> Please verify your email address or your domain may be suspended.

It **will** be suspended if not verified (in 2 weeks), and suspension means the site stops working. `Domain may be suspended` doesn't imply that, IMO.

From our current messaging, it isn't clear what the consequences are of suspended domain, and when a site does stop working, some don't clearly know "why".

ranh commented:
> It will be suspended if not verified (in 2 weeks), and suspension means the site stops working

@ehti how about "your domain will stop working" instead of "your domain will be suspended"?

ehti commented:
That sounds good @ranh .

@klimeryk do we show the verification warning for ccTLDs or other TLDs too that don't require verification?

klimeryk commented:
I like @folletto's suggestion to make the resend action less prominent, the inline link seems like a good option, that should address @ranh's concerns?

While we should keep the copy fairly short and to-the-point, I think @ehti's suggestion to explicitly mention checking spam/junk should be included. It might be obvious to us, but experience shows that users don't do that ;/ I like the `Is the email incorrect? Change it from here.` part - that might prompt the user to double-check the email.
We also know what's the email associated with the user's account. From what I've seen, a lot of users are confused about the WPCOM email vs WHOIS email deal. Maybe we could use that info to further improve the chances the user checks the correct inbox (or updates the WHOIS email if it's not in sync with the WPCOM one). Ideally, the WPCOM one and the WHOIS one would be bound together, but for different reasons, I can imagine some users not wanting that (if they don't have privacy enabled, the WHOIS info is public, so they might want to use an alias, etc.).

klimeryk commented:
> @klimeryk do we show the verification warning for ccTLDs or other TLDs too that don't require verification?

No, we should not be - if it's `verified` in DARC, it should not show a notice in Calypso. I've fixed that for ccTLDs some time ago - please let me know if that's not the case.

mikeshelton1503 commented:
Agree the Send button was a bit too prominent. I switched it up a bit to prioritize the "Sent to [email address" messaging as @folletto and @ehti suggested.

Using compact buttons for both actions.

![screen shot 2017-07-10 at 11 28 41 am](https://user-images.githubusercontent.com/12596797/28026129-2531c190-6563-11e7-8318-dffb3a8d5ea8.png)

>From our current messaging, it isn't clear what the consequences are of suspended domain, and when a site does stop working, some don't clearly know "why".

Updated copy per @ranh suggested copy.

folletto commented:
Looks good to me! Nice work! 💯

mikeshelton1503 commented:
The only thing I'm not sure about now is when you send the email, we show the sent message without the action buttons. The problem is if you didn't receive the email after that, you might look for the action buttons again but it's not clear how to get them back (refresh or navigate way).

@folletto @breezyskies what do you think about fading that message out after maybe 5 seconds. Alternatively could add an "X" or close action but that feel unnecessary.

folletto commented:
Sorry for asking, but can you show the sequence of mockups in that scenario? I think I got it, but I feel I'm doing too many mental jumps.

mikeshelton1503 commented:
@folletto I just updated the PR with screenshots and test instructions. Does that help?

ranh commented:
@mikeshelton1503 How about this for the description?

> We need to check your contact information to make sure you can be reached. Please verify your details using the email we sent you, or your domain will stop working. 

Also, where it currently says "verification sent to mail@example.com", that's a bit unclear. How about "instructions sent to mail@example.com"?

mikeshelton1503 commented:
> How about this for the description?

👍 

>Also, where it currently says "verification sent to mail@example.com", that's a bit unclear. How about "instructions sent to mail@example.com"?

Yea I wondered about that phrasing. Though not sure if instructions is quite right either - it implies that there are steps to it. What about one of these:

- Check your email — sent to mike.shelton@automattic.com.
- Check your email — we sent it to mike.shelton@automattic.com.
- Check your email — verification link sent to mike.shelton@automattic.com.
- Check your email — link to verify sent to mike.shelton@automattic.com.

ranh commented:
> it implies that there are steps to it

That's intentional: after clicking the button in the email, you still have to click another button in the screen that loads up.

mikeshelton1503 commented:
@klimeryk Could you put the "Sent" message on a timer so that it disappears after 5 seconds? This is the last thing to be done before this is officially ready for review.

klimeryk commented:
@mikeshelton1503 Sure thing - timer added :)

mikeshelton1503 commented:
This is ready for review! cc @breezyskies @klimeryk

breezyskies commented:
Design 👍

klimeryk commented:
I've pushed one more small fix (nothing impacting design/logic, just developer stuff) and rebased with master because in the meantime there's been a pretty important update to webpack 3 which might break some stuff (though, that's just me being overly cautious).

klimeryk commented:
And one more small fix in case the user has a long email, but a small viewport ;)
Before:
<img width="404" alt="screen shot 2017-07-12 at 14 51 44" src="https://user-images.githubusercontent.com/3392497/28118563-cc0b6c0a-6711-11e7-8f59-2418e9a19f3f.png">

After:
<img width="395" alt="screen shot 2017-07-12 at 14 54 14" src="https://user-images.githubusercontent.com/3392497/28118609-006e546c-6712-11e7-89cc-6e8dd9768ad9.png">

klimeryk commented:
@hewsut No worries, I've noticed that it's not used anywhere ATM, so of course some regression might creep in. It was just funny to see that it's not actually working after all that feedback and code reviews - rule #1 of code reviews: actually run the code 😜

aidvu commented on client/components/data/query-whois/index.jsx:
```diff
@@ -30,6 +30,6 @@ QueryWhois.propTypes = {
 };
 
 export default connect(
-	( state, { domain } ) => ( { requesting: isRequestingWhois( state, domain ) } ),
```
Ow wow. Good job catching this.

aidvu commented on client/my-sites/domains/domain-management/components/icann-verification/icann-verification-card.jsx:
```diff
@@ -2,61 +2,164 @@
+	( state, ownProps ) => ( {
+		contactDetails: getRegistrantWhois( state, ownProps.selectedDomainName ),
+	} ),
+	dispatch => bindActionCreators( { errorNotice, successNotice }, dispatch )
```
I don't think you need to do this anymore. If you just pass
```
{ errorNotice, successNotice }
``` 
it will automagically bind: https://github.com/Automattic/wp-calypso/pull/15058/files#diff-9b3dc846fe889e4ab5d78da8b44490ddR208

klimeryk replied:
TIL, thanks! Also, while fixing this I remembered that `successNotice` is no longer used and can be removed 😀

aidvu commented on client/my-sites/domains/domain-management/components/icann-verification/icann-verification-card.jsx:
```diff
@@ -2,61 +2,164 @@
+
+		this.setState( { submitting: true } );
+
+		upgradesActions.resendIcannVerification( this.props.selectedDomainName, ( error ) => {
```
Would be amazing if we move this to `redux` as part of the refactor. Your call.

klimeryk replied:
I agree, but let's not do this in this PR. The discussion has been long enough and this refactor and subsequent testing would block this for another few days and I don't want to block Mike for that long. It might be a good idea for a mob programming session, though :)
